From ded4c3b690446edbeef5f83e3a24493d92eba062 Mon Sep 17 00:00:00 2001
From: Kevin Chabot <Kevin_Chabot@hotmail.com>
Date: Tue, 13 Dec 2022 19:25:51 -0500
Subject: [PATCH] Server: Synchronise keystate when locking input when the
 hardware message is queued.

Fix an issue where modifier keys stay stuck when using a drawing tablet
and a keyboard using Clip Studio Paint.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=33056
---
 server/queue.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/server/queue.c b/server/queue.c
index 11111111111..11111111111 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -703,7 +703,11 @@ static inline void set_queue_bits( struct msg_queue *queue, unsigned int bits )
 {
     if (bits & (QS_KEY | QS_MOUSEBUTTON))
     {
-        if (!queue->keystate_lock) lock_input_keystate( queue->input );
+        if (!queue->keystate_lock)
+        {
+            sync_input_keystate( queue->input );
+            lock_input_keystate( queue->input );
+        }
         queue->keystate_lock = 1;
     }
     queue->wake_bits |= bits;
-- 
0.0.0

